scylla-cql/src/types/deserialize/frame_slice.rs

```diff
@@ -0,0 +1,188 @@
+use bytes::Bytes;
+
+use crate::frame::frame_errors::ParseError;
+use crate::frame::types;
+
+/// A reference to a part of the frame.
+//
+// # Design justification
+//
+// ## Why we need a borrowed type
+//
+// The reason why we need to store a borrowed slice is that we want a lifetime that is longer than one obtained
+// when coercing Bytes to a slice in the body of a function. That is, we want to allow deserializing types
+// that borrow from the frame, which resides in QueryResult.
+// Consider a function with the signature:
+//
+// fn fun(b: Bytes) { ... }
+//
+// This function cannot return a type that borrows from the frame, because any slice created from `b`
+// inside `fun` cannot escape `fun`.
+// Conversely, if a function has signature:
+//
+// fn fun(s: &'frame [u8]) { ... }
+//
+// then it can happily return types with lifetime 'frame.
+//
+// ## Why we need the full frame
+//
+// We don't. We only need to be able to return Bytes encompassing our subslice. However, the design choice
+// was made to only store a reference to the original Bytes object residing in QueryResult, so that we avoid
+// cloning Bytes when performing subslicing on FrameSlice. We delay the Bytes cloning, normally a moderately
+// expensive operation involving cloning an Arc, up until it is really needed.
+//
+// ## Why not different design
+//
+//      - why not a &'frame [u8] only? Because we want to enable deserializing types containing owned Bytes, too.
+//      - why not a Bytes only? Because we need to propagate the 'frame lifetime.
+//      - why not a &'frame Bytes only? Because we want to somehow represent subslices, and subslicing
+//        &'frame Bytes return Bytes, not &'frame Bytes.
+#[derive(Clone, Copy, Debug)]
```

Comment:
Another option worth mentioning here would be to use unsafe: you could store only Bytes and transmute the lifetime of the returned slice.

Reply:
I think it is agreed upon that we don't allow any `unsafe` code in the driver itself.
Btw, we could enforce it by using `#![forbid(unsafe_code)` crate-wide. WDYT?

Reply:
I know we don't use unsafe, that's why I only want to mention it as an alternative, not implement it.

We can enforce it, but I don't think it's important, I doubt any unsafe can get through review unnoticed.